Document guideline to design TruffleLanguage with language-agnostic initialization.

smarr opened the pull request:
A TruffleLanguage should provide a language-agnostic interface to be able to be usable in multi-language environments. Language-specific front-ends or launchers, for instance, emulating existing implementations, should be realized externally to the PolyglotEngine.

This change fixes also minor issues in existing comments.

The tests on Travis still fail, because it seems to use the Eclipse 4.5 formatter.
Locally, it passes with using Eclipse 4.4 and making sure `ECLIPSE_EXE` is set.

Recreated the PR. Seems like force-push to a branch after the original PR is closed is problematic.

jtulach commented:
The formatting looks OK.

jtulach commented on truffle/com.oracle.truffle.api.vm/src/com/oracle/truffle/api/vm/PolyglotEngine.java:
```diff
@@ -80,18 +80,26 @@
+ * A {@link com.oracle.truffle.api.TruffleLanguage} should be designed to be initializable and
+ * configurable in a generic way. Language-specific execution front-ends, for instance to emulate
+ * existing language implementations, should be realized externally to maintain a consistent and
+ * language-agnostic way of initializing {@link PolyglotEngine}.
```
@chumer: "I am not sure this should be here. I think having it on the TruffleLanguage should be good enough.
From a PolyglotEngine user / integrator perspective this is confusing."

smarr replied:
@chumer My goal is to make sure people reading about `PolyglotEngine` also know about the design guidelines for `TruffleLanguage` in order to understand what one can expect from them. IMHO, redundancy is a good thing for that goal.
But perhaps, we can adapt the wording to fit better within the context here?

chumer replied:
The design guidelines should be described in detail in TruffleLanguage. I'd rather do a @see reference to TruffleLanguage on how to implement the language. It just looks weird to mix SPI and API documentation like this.

jtulach replied:
It would be better if the documentation of `PolyglotEngine` was more focused on users of the languages. For example:

```
If the way you use a language requires special setup,
you should obtain necessary parameters from somewhere (for example 
when parsing a command line) and communicate them to
the {@link PolyglotEngine} - either via {@link PolyglotEngine.Builder#config} 
as exposed by the language implementation or by evaluating 
appropriate "prelude" scripts via {@link PolyglotEngine.Language#eval}.
Possibly you can pre-register various {@link PolyglotEngine.Builder#globalSymbol global objects}
and make them available to the system of {@link PolyglotEngine}.
```

chumer replied:
@jtulach  Looks good. I am fine with that.

smarr replied:
ok, @jtulach are you changing that directly, or do you want me to update the PR?

jtulach replied:
Stefan, please update your PR. I like your language changes (already part of the change), so I am looking forward to see how you polish my Czenglish in this paragraph.

chumer replied:
Its going to be a mixture between Czenglish and Denglish. Nice! :D

smarr commented:
I incorporated @jtulach's text, albeit slightly reworded and reordered, putting the emphasize and the ways to configure the engine/language.

PR is also rebased on current master.

mlvdv commented:
Looks good; thanks.
